[Flang] Remove FLANG_INCLUDE_RUNTIME

Meinersbur opened the pull request:
Remove the FLANG_INCLUDE_RUNTIME option which was replaced by LLVM_ENABLE_RUNTIMES=flang-rt.

The FLANG_INCLUDE_RUNTIME option was added in #122336 which disables the non-runtimes build instructions for the Flang runtime so they do not conflict with the LLVM_ENABLE_RUNTIMES=flang-rt option added in #110217. In order to not maintain multiple build instructions for the same thing, this PR completely removes the old build instructions (effectively forcing FLANG_INCLUDE_RUNTIME=OFF). 

As per discussion in https://discourse.llvm.org/t/buildbot-changes-with-llvm-enable-runtimes-flang-rt/83571/2 we now implicitly add LLVM_ENABLE_RUNTIMES=flang-rt whenever Flang is compiled in a bootstrapping (non-standalone) build. Because it is possible to build Flang-RT separately, this behavior can be disabled using `-DFLANG_ENABLE_FLANG_RT=OFF`. Also see the discussion an implicitly adding runtimes/projects in #123964.

This PR will remains in draft status until the buildbots' configurations have been updated. Due to LLVM_ENABLE_RUNTIMES=flang-rt being added automatically, most of them would probably continue to work, except the out-of-tree builders which require some adjustment.

Series of buildbot updates:
 * https://github.com/llvm/llvm-zorg/pull/383
 * https://github.com/llvm/llvm-zorg/pull/386
 * https://github.com/llvm/llvm-zorg/pull/387
 * https://github.com/llvm/llvm-zorg/pull/389
 * https://github.com/llvm/llvm-zorg/pull/391
 * https://github.com/llvm/llvm-zorg/pull/401
 * https://github.com/llvm/llvm-zorg/pull/393
 * https://github.com/llvm/llvm-zorg/pull/396
 * https://github.com/llvm/llvm-project/pull/128678
 * https://github.com/llvm/llvm-zorg/pull/402
 * https://github.com/llvm/llvm-project/pull/129692
 * https://github.com/llvm/llvm-zorg/pull/388 (optional)

vzakhari commented:
Hi @Meinersbur is this the only patch that is left to be merged?  Could you please rebase it so that I can try it on top-of-tree?

Meinersbur committed:
Remove FLANG_INCLUDE_RUNTIME

Meinersbur committed:
Implicitly enable Flang-RT

Meinersbur commented:
Rebased and added implicitly adding `LLVM_ENABLE_RUNTIMES=flang-rt` according to https://discourse.llvm.org/t/buildbot-changes-with-llvm-enable-runtimes-flang-rt/83571/2

vzakhari commented:
Thank you for the rebase, @Meinersbur.  Please give me some time to adjust build tools on my side to make it work with `flang/runtime` removed.  I will let you know, when I am ready for the merge.

Meinersbur committed:
Merge branch 'main' into users/meinersbur/flang_runtime_remove-FLANG_INCLUDE_RUNTIME

Meinersbur commented:
All the buildbots and the pre-merge CI have been updated. Ready for review.

jhuber6 commented on llvm/CMakeLists.txt:
```diff
@@ -187,6 +187,19 @@ foreach(proj IN LISTS LLVM_ENABLE_RUNTIMES)
+    message(STATUS "Enabling Flang-RT as a dependency of Flang"
+      "Flang-RT is required for a fully working Flang toolchain."
+      "If you intend to compile Flang-RT separately following the instructions at"
+      "https://github.com/llvm/llvm-project/blob/main/flang-rt/README.md#standalone-runtimes-build"
```
Directly linking to a github readme is a little weird. Do we not have a documentation page for this?

Meinersbur replied:
There is no flang-rt.llvm.org . I do find several links to github.com in llvm/docs, such as https://github.com/llvm/llvm-project/blob/9387281b6e989832d01db352e6e9068d0b4ce8c2/llvm/docs/TableGen/index.rst?plain=1#L291

jhuber6 replied:
There's https://flang.llvm.org/docs/ right?

Meinersbur replied:
Is it ok to update the docs in a follow-up? Consolidating this README.md and https://flang.llvm.org/docs/GettingStarted.html#building-flang-rt-for-accelerators itself would be a larger patch. Also think that advanced topics like FLANG_RT_EXPERIMENTAL_OFFLOAD_SUPPORT does not belong into a GettingStarted.

Meinersbur replied:
I moved the help `message(...)` to flang/ so it is also emitted in a flang-standalone build. I also shortened it without the README link, may update it after consolidation.

jhuber6 commented on flang/examples/ExternalHelloWorld/CMakeLists.txt:
```diff
@@ -1,9 +0,0 @@
-# This test is not run by default as it requires input.
```
Guessing this was already copied somewhere else?

Meinersbur replied:
flang-rt version of this example is here: https://github.com/llvm/llvm-project/blob/main/flang-rt/examples/ExternalHelloWorld/CMakeLists.txt

This is an example that links to flang-rt from C++. No use of a Fortran compiler. I think I should delete the .cpp as well.

vzakhari commented:
Everything is ready on my side.  Thank you for waiting, Michael! LGTM

Meinersbur committed:
Remove dead CMakeLists.txt

Meinersbur committed:
Move 'no runtime' warning to flang

Meinersbur committed:
Fix syntax error

Meinersbur committed:
Merge branch 'main' into users/meinersbur/flang_runtime_remove-FLANG_INCLUDE_RUNTIME

Meinersbur commented:
After AMD gave their OK as well, I am going to merge this.

Meinersbur committed:
Merge branch 'main' into users/meinersbur/flang_runtime_remove-FLANG_INCLUDE_RUNTIME

Meinersbur committed:
Group implicitly enabled projects/runtimes together

Meinersbur commented:
The two builders failing were supposed to be updated by https://github.com/llvm/llvm-zorg/pull/393. Have to wait until https://lab.llvm.org/buildbot/ is restarted before relanding.

DavidTruby commented:
This patch seems to have removed the behaviour that flang-rt is added to LLVM_ENABLE_RUNTIMES automatically which has broken all of our internal builds. My read of the commit message, and the patch, is that this behaviour hasn't been intentionally removed?
I have a cmake line like:
```
cmake -GNinja -DCMAKE_C_COMPILER=clang -DCMAKE_CXX_COMPILER=clang++ -DLLVM_CCACHE_BUILD=On -DLLVM_ENABLE_PROJECTS="clang;flang;mlir" -DLLVM_ENABLE_RUNTIMES=openmp  -DCMAKE_BUILD_TYPE=Release -DLLVM_ENABLE_ASSERTIONS=On ../llvm
```
and I am not getting flang-rt. I also see:
```
-- Not building Flang-RT. For a usable Fortran toolchain, either set FLANG_ENABLE_FLANG_RT=ON, add LLVM_ENABLE_RUNTIMES=flang-rt, or compile a standalone Flang-RT.
```

DavidTruby commented on llvm/CMakeLists.txt:
```diff
@@ -149,6 +149,12 @@ if ("flang" IN_LIST LLVM_ENABLE_PROJECTS)
+  option(FLANG_ENABLE_FLANG_RT "Implicitly add LLVM_ENABLE_RUNTIMES=flang-rt when compiling Flang" ON)
+  if (FLANG_ENABLE_FLANG_RT AND NOT "flang-rt" IN_LIST LLVM_ENABLE_RUNTIMES)
+    message(STATUS "Enabling Flang-RT as a dependency of Flang")
+    list(APPEND LLVM_ENABLE_RUNTIMES "flang-rt")
```
I believe you can't `list(APPEND ..)` cache variables like this as they'll get overridden by what was provided on the command line. I think it should be `set(LLVM_ENABLE_RUNTIMES "${LLVM_ENABLE_RUNTIMES};flang-rt" CACHE INTERNAL "")` or something along those lines

DavidTruby replied:
Hmm, that doesn't seem to work either. I actually don't see the message so for some reason we aren't entering that if statement body I guess?

DavidTruby replied:
Sorry. All of that was false alarms. The issue is that LLVM_ENABLE_RUNTIMES hasn't been defined by this point, it's defined a little after these checks. When it does get defined it overwrites whatever was done here. 

If you move the LLVM_ENABLE_RUNTIMES define above this, then this works as expected.

DavidTruby commented:
I have posted #138136 which fixes the issues I mentioned above
